[#5472] improvement(docs): Add example to use cloud stroage fileset and polish hadoop-catalog document. (#6059)
apache/gravitino: docs/cloud-storage-fileset-example.md

---
title: "How to use cloud storage fileset"
slug: /how-to-use-cloud-storage-fileset
keyword: fileset S3 GCS ADLS OSS
license: "This software is licensed under the Apache License version 2."
---

> This document aims to provide a comprehensive guide on how to use cloud storage fileset created by Gravitino, it usually contains the following sections:

Reviewer: End a sentence with a full period `.`; start a new sentence with a capital letter,
and optionally leave two spaces between two sentences.

| S3           | AWS S3.                                                       | [gravitino-aws-hadoop-bundle](https://mvnrepository.com/artifact/org.apache.gravitino/gravitino-hadoop-aws-bundle)       | 0.8.0-incubating |
| GCS          | Google Cloud Storage.                                         | [gravitino-gcp-hadoop-bundle](https://mvnrepository.com/artifact/org.apache.gravitino/gravitino-hadoop-gcp-bundle)       | 0.8.0-incubating |
| OSS          | Aliyun OSS.                                                   | [gravitino-aliyun-hadoop-bundle](https://mvnrepository.com/artifact/org.apache.gravitino/gravitino-hadoop-aliyun-bundle) | 0.8.0-incubating |
| ABS          | Azure Blob Storage (aka. ABS, or Azure Data Lake Storage (v2) | [gravitino-azure-hadoop-bundle](https://mvnrepository.com/artifact/org.apache.gravitino/gravitino-hadoop-azure-bundle)   | 0.8.0-incubating |

After adding the jars into the fileset catalog classpath, you can start up the Gravitino server by running the following command:

```shell
cd ${GRAVITINO_HOME}
bin/gravitino.sh start
```

### Bundle jars

> Gravitino bundles jars are jars that are used to access the cloud storage, they are divided into two categories:

Reviewer: We have two sentences here... and the wordy statement can be simplified without losing anything:

```suggestion
Gravitino bundles jars are used to access the cloud storage.
They are divided into two categories:
```

- `gravitino-${aws,gcp,aliyun,azure}-bundle-{gravitino-version}.jar` are the jars that contain all the necessary dependencies to access the corresponding cloud storages. For instance, `gravitino-aws-bundle-${gravitino-version}.jar` contains the all necessary classes including `hadoop-common`(hadoop-3.3.1) and `hadoop-aws` to access the S3 storage.

> They are used in the scenario where there is no hadoop environment in the runtime.

Reviewer: Always follow the existing spelling of established words, e.g. Hadoop, Apache, AWS.

```suggestion
They are used when there is no Hadoop environment in the runtime.
```

|------------------------|------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------|
| No Hadoop environment  | `gravitino-aws-bundle-${gravitino-version}.jar`                                                                                          | `gravitino-gcp-bundle-${gravitino-version}.jar`                                                              | `gravitino-aliyun-bundle-${gravitino-version}.jar`                                                                                         | `gravitino-azure-bundle-${gravitino-version}.jar`                                                                  |

> | 2.x, 3.x               | `gravitino-aws-${gravitino-version}.jar`, `hadoop-aws-${hadoop-version}.jar`, `aws-sdk-java-${version}` and other necessary dependencies | `gravitino-gcp-{gravitino-version}.jar`, `gcs-connector-${hadoop-version}`.jar, other necessary dependencies | `gravitino-aliyun-{gravitino-version}.jar`, hadoop-aliyun-{hadoop-version}.jar, aliyun-sdk-java-{version} and other necessary dependencies | `gravitino-azure-${gravitino-version}.jar`, `hadoop-azure-${hadoop-version}.jar`, and other necessary dependencies |

Reviewer: We don't have to use tables here. Markdown sucks when used to create tables, especially when the cell contents are long.
We can use, for example, unordered lists for this.

For `hadoop-aws-${hadoop-version}.jar`, `hadoop-azure-${hadoop-version}.jar` and `hadoop-aliyun-${hadoop-version}.jar` and related dependencies, you can get them from `${HADOOP_HOME}/share/hadoop/tools/lib/` directory.
For `gcs-connector`, you can download it from the [GCS connector](https://github.com/GoogleCloudDataproc/hadoop-connectors/releases) for hadoop2 or hadoop3. 

> If there still have some issues, please report it to the Gravitino community and create an issue. 

Reviewer: Let's make the users' life easier by providing a link.
Or else they have to check where is the community and/or where is the repo.

```suggestion
If there are still issues, please consider [filing an issue](...) 
```

    .withMetalake("metalake")
    .build();

gcsProperties = ImmutableMap.<String, String>builder()
    .put("location", "gs://bucket/root")
    .put("gcs-service-account-file", "path_of_gcs_service_account_file")
    .put("filesystem-providers", "gcs")
    .build();

Catalog gcsCatalog = gravitinoClient.createCatalog("catalog",
    Type.FILESET,
    "hadoop", // provider, Gravitino only supports "hadoop" for now.
    "This is a GCS fileset catalog",
    gcsProperties);
// ...

```

</TabItem>
<TabItem value="python" label="Python">

```python
gravitino_client: GravitinoClient = GravitinoClient(uri="http://localhost:8090", metalake_name="metalake")

gcs_properties = {
    "location": "gcs://bucket/root",
    "gcs_service_account_file": "path_of_gcs_service_account_file"
}

s3_catalog = gravitino_client.create_catalog(name="catalog",
                                             type=Catalog.Type.FILESET,
                                             provider="hadoop",
                                             comment="This is a GCS fileset catalog",
                                             properties=gcs_properties)

```

</TabItem>
</Tabs>

:::note

> The prefix of a GCS location should always start with `gs` for instance, `gs://bucket/root`.

Reviewer: ```suggestion
The prefix of a GCS location should always start with `gs`, for instance, `gs://bucket/root`.
```

<TabItem value="shell" label="Shell">

```shell
curl -X POST -H "Accept: application/vnd.gravitino.v1+json" \
-H "Content-Type: application/json" -d '{
  "name": "schema",
  "comment": "comment",
  "properties": {
    "location": "file:///tmp/root/schema"
  }
}' http://localhost:8090/api/metalakes/metalake/catalogs/catalog/schemas
```

</TabItem>
<TabItem value="java" label="Java">

```java
GravitinoClient gravitinoClient = GravitinoClient
    .builder("http://localhost:8090")
    .withMetalake("metalake")
    .build();

// Assuming you have just created a Hadoop catalog named `catalog`
Catalog catalog = gravitinoClient.loadCatalog("catalog");

SupportsSchemas supportsSchemas = catalog.asSchemas();

Map<String, String> schemaProperties = ImmutableMap.<String, String>builder()
    // Property "location" is optional, if specified all the managed fileset without
    // specifying storage location will be stored under this location.
    .put("location", "file:///tmp/root/schema")
    .build();
Schema schema = supportsSchemas.createSchema("schema",
    "This is a schema",
    schemaProperties
);
// ...
```

</TabItem>
<TabItem value="python" label="Python">

> You can change the value of property `location` according to which catalog you are using, moreover, if we have set the `location` property in the catalog, we can omit the `location` property in the schema.

Reviewer: ```suggestion
You can change the `location` value based on the catalog you are using.
If the `location` property is specified in the catalog, we can omit it in the schema.
```

spark_df.coalesce(1).write
  .mode("overwrite")
  .option("header", "true")
  .csv(gvfs_path)

```

If your Spark without Hadoop environment, you can use the following code snippet to access the fileset:

```python
## replace the env PYSPARK_SUBMIT_ARGS variable in the code above with the following content:
### S3
os.environ["PYSPARK_SUBMIT_ARGS"] = "--jars /path/to/gravitino-aws-bundle-{gravitino-version}.jar,/path/to/gravitino-filesystem-hadoop3-runtime-{gravitino-version}.jar --master local[1] pyspark-shell"
### GCS
os.environ["PYSPARK_SUBMIT_ARGS"] = "--jars /path/to/gravitino-gcp-bundle-{gravitino-version}.jar,/path/to/gravitino-filesystem-hadoop3-runtime-{gravitino-version}.jar, --master local[1] pyspark-shell"
### OSS
os.environ["PYSPARK_SUBMIT_ARGS"] = "--jars /path/to/gravitino-aliyun-bundle-{gravitino-version}.jar,/path/to/gravitino-filesystem-hadoop3-runtime-{gravitino-version}.jar, --master local[1] pyspark-shell"
#### Azure Blob Storage
os.environ["PYSPARK_SUBMIT_ARGS"] = "--jars /path/to/gravitino-azure-bundle-{gravitino-version}.jar,/path/to/gravitino-filesystem-hadoop3-runtime-{gravitino-version}.jar --master local[1] pyspark-shell"
```

:::note

> **In some Spark version, Hadoop environment is needed by the driver, adding the bundle jars with '--jars' may not work, in this case, you should add the jars to the spark classpath directly.**

Reviewer: You don't need to use bold fonts here since the text is already in a `note`.
